### PR TITLE
add optional query metadata

### PR DESCRIPTION
**Pull request by m-burgoyne**
Updated the collection schema to add a property object to support custom metadata for individual query types. The implementation is different than the original proposal in that it has named properties for the queries which require custom information rather than an array of objects, the aim of this approach is to simplify looking up metadata required for a specific query. Details based on the ideas described in issue #166.

### CODE REVIEW

**Review by tomkralidis — CHANGES_REQUESTED** [review]
Can we addd an example collection response in `candidate-standard/examples/collection.json` type thing?  This will also help visualize this PR a bit better (for me).

**Comment by m-burgoyne** [comment]
@tomkralidis example added

**Comment by tomkralidis** [comment]
Here's the thinking in OARec at the moment about adding binding information to collections: https://github.com/opengeospatial/ogcapi-records/issues/95#issuecomment-785087799

This means adding a `parameters` block to collection metadata.  However, `parameters` may be confusing in our context (i.e. `parameter-name`, etc.)?

**Review by chris-little — APPROVED** [review]
That was easy. Don't forget the software implementations! ;-)

**Comment by iandruska-ibl** [comment]
I wonder why CRS is described differently at the collection level vs data query level.

`collection.yaml` - just a string:
```
  crs:
    description: list of the coordinate reference systems the collection results can support
    type: array
    items:
      type: string
```

`dataQuery.yaml` - a CRS object:
```
crs:
    description: List of key/value definitions for the CRS's supported by a query.  The key is the query parameter and the value is the Well Known Text description
    type: array
    items:
      $ref: crsObject.yaml
```

P.S. The description at the data query level indicates that the type is object but is actually array.

**Comment by iandruska-ibl** [comment]
@chris-little Could you please make me a reviewer instead of Pavol?

**Comment by iandruska-ibl** [comment]
Our property names use three different case styles:

1. camelCase - `dataQueries`, `defaultOutputFormat`, ...
2. kebab-case - `parameter-name`, `width-units`, `height-units`
3. nocase - `outputformat`

I believe we should use case styles consistently. I was not able to find any OGC regulation on this. However, our OGC-API Common colleagues use camelCase consistently in their schemas.

**Comment by m-burgoyne** [comment]
@iandruska-ibl I have updated the proposed changes to make the `crs` definitions more consistent between the query level metadata and collection level metadata.

**Comment by iandruska-ibl** [comment]
> @iandruska-ibl I have updated the proposed changes to make the `crs` definitions more consistent between the query level metadata and collection level metadata.

@m-burgoyne I do not see the changes. The `crs` attribute at collection level is still an array of strings while on data query level it is an array of `crsObject.yaml`.

**Comment by m-burgoyne** [comment]
Hi @iandruska-ibl the data-queries object now has an attribute of `crs-detail`which consists of an array of crsObject.yaml.  The crsObject.yaml has been modified to be an object with attributes of `crs` as a string (to match the collection definition) and `wkt` which is the Well Known Text definition of the crs value

**Comment by iandruska-ibl** [comment]
@m-burgoyne So the collection's `crs` attribute is just a reference to the actual CRS definition inside data query, is that correct? 

However, I am not sure what problem does this arrangement solve. I would expect the opposite arrangement: definition on the collection level and just reference on the data query level. It would eliminate duplicate CRS definitions.

**Comment by m-burgoyne** [comment]
Hi @iandruska-ibl this approach means that the schema at collection level remains consistent with the OGC API Features schema whilst allowing the EDR API a mechanism to provide a WKT definition of the CRS and a name which can be used as a query parameter value.

File: candidate-standard/openapi/schemas/collection.yaml [hunk 1]
```diff
@@ -58,24 +58,23 @@ properties:
...
-  distanceunits:
-    description: list of distance units radius distance values can be specified in
-    type: array
-    items: 
       type: string
-    example:
-      - km
-      - miles
-  outputformat:
+  data-queries:
```
Comment (iandruska-ibl) [inline candidate-standard/openapi/schemas/collection.yaml:65]: This is not a valid OpenAPI. I believe it should be
```
  data-queries:
    description: Detailed information relevant to individual query types
    type: object
    properties:
      position:
        $ref: dataQuery.yaml
      radius:
        $ref: dataQuery.yaml
      area:
        $ref: dataQuery.yaml
      cube:
        $ref: dataQuery.yaml
      trajectory:
        $ref: dataQuery.yaml
      corridor:
        $ref: dataQuery.yaml
 ```

**Review by iandruska-ibl — CHANGES_REQUESTED** [review]
Collection schema needs a fix.

File: candidate-standard/openapi/schemas/collection.yaml (hunk 1 above)
Comment (m-burgoyne) [reply candidate-standard/openapi/schemas/collection.yaml:65]: Thanks @iandruska-ibl, I have uploaded now upload a correction

**Comment by chris-little** [comment]
I think our proposed changes are consistent with the draft OGC API-Common Part 1 Core which is now going to public comment (we can comment if not), but are we consistent with the latest version of OGC API-Common Part 2: Collections (if we ignore the renaming of collections to GeoData or something similar) as [presented to the Coverages SWG](https://portal.ogc.org/files/?artifact_id=96817) ?

**Comment by tomkralidis** [comment]
Update on current thinking in OARec: each association link would have an OPTIONAL `variables` object with binding options specific to URI template specified in the `href`.  See https://github.com/opengeospatial/ogcapi-records/blob/master/core/examples/json/record.json#L77 for an example.

**Review by iandruska-ibl — APPROVED** [review]

**Review by chris-little — APPROVED** [review]

**Comment by tomkralidis** [comment]
Pending discussion in https://github.com/opengeospatial/ogcapi-environmental-data-retrieval/issues/233#issuecomment-808241124

**Comment by iandruska-ibl** [comment]
I just hit a snag trying to import EDR Open API specification into AWS API Gateway. Turned out that the API Gateway does not accept the kebab case attribute names in our schema (`crs-details`, `within-units`, `parameter-names`, ...). It accepts only `[a-zA-Z0-9]+` attributes. Actually, specification can be imported, but only resources (endpoints) are created. Response schemas are completely ignored.

I realise that AWS is not something we are bound to obey, but on the other hand AWS is too big to be ignored. And I think I agree with AWS that kebab-case attribute names are a bit unorthodox. Should we consider changing to camel case?

@m-burgoyne @tomkralidis @chris-little

**Comment by tomkralidis** [comment]
We should definitely table with other API SWGs.  Out of interest, does snake_case work?  Note that we also have kebab-case in OARec.  To clarify, this applies ONLY to the schema (and not the endpoint query parameters), correct?

**Comment by iandruska-ibl** [comment]
snake_case seems to work.

Yes, this applies only to schema, not to URL parameters.

This is an old issue API Gateway developers are aware of, but they seem to ignore it: https://forums.aws.amazon.com/thread.jspa?threadID=233021

**Comment by chris-little** [comment]
So, we can use `lower-kebab-case` for the OpenAPI V3 parameters, as they appear in the URL, but for OpenAPI V3 Schema components, we can use `lower_snake_case`?

**Comment by tomkralidis** [comment]
I asked in the [STAC Gitter channel](https://gitter.im/SpatioTemporal-Asset-Catalog/Lobby?at=605e1497559d0e099a60d8d3) who provided great feedback.

Proposal, then:
- kebab-case for endpoints and query parameters
- snake_case for schema properties

Thoughts?  I'm happy to make an update/PR on this one if folks agree.

**Comment by chris-little** [comment]
I support a PR, I guess you have a week or so beofre the next meeting. Let us see what @m-burgoyne @iandruska-ibl @ShaneMill1 @boyi-shangguan think. Mark will be on holiday next week and after too.

**Comment by solson-nws** [comment]
I support a PR as well.  Makes sense to me.

Steve

**Comment by tomkralidis** [comment]
I've opened this as an [issue](https://github.com/opengeospatial/ogcapi-common/issues/224) in OGC API - Common as I think this will apply / help other APIs.

**Comment by ghobona** [comment]
A similar conversation is taking place at https://github.com/opengeospatial/ogcapi-records/issues/108

**Comment by m-burgoyne** [comment]
@chris-little @tomkralidis @iandruska-ibl @ShaneMill1 @boyi-shangguan @solson-nws I agree with the suggested approach and I have updated the pull request to replace the kebab case entries in the yaml schema properties with snake case values. I have also updated the schemas to move the optional data query properties to the a new variables property in the links section (as discussed in SWG 48)

**Review by tomkralidis — APPROVED** [review]
Note the discussion in [OGC API - Common](https://github.com/opengeospatial/ogcapi-common/issues/224). Having said this, either snake_case or lowerCamelCase is safe.

**Review by iandruska-ibl — APPROVED** [review]

File: candidate-standard/openapi/examples/collection.json [hunk 2]
```diff
@@ -0,0 +1,625 @@
...
+                    "title": ""
+                },
+                {
+                    "href": "https://www.metoffice.gov.uk/services/data/datapoint/terms-and-conditions---datapoint#termsofservice",
+                    "hreflang": "en",
+                    "rel": "restrictions",
+                    "type": "text/html",
+                    "title": ""
+                },
+                {
```
Comment (iandruska-ibl) [inline candidate-standard/openapi/examples/collection.json:61]: What indicates that this object describes a data query? I believe that "type" attribute is misused here. It is supposed to be a Media/Content Type.

**Comment by chris-little** [comment]
@m-burgoyne @tomkralidis @iandruska-ibl Can we quickly address this lasts issue from @iandruska-ibl, then we can merge so we can prepare a final version of the standard for publication, with whatever name we end up with, and assuming the planning Committee says yes, and does not delay us.

**Comment by m-burgoyne** [comment]
@chris-little @iandruska-ibl @tomkralidis The type attribute is optional in the `link` schema so I have remove that attribute from the data query examples.

**Comment by iandruska-ibl** [comment]
@m-burgoyne But a software client still cannot tell whether the link describes/points to a data query. I think that the previous solution (`data_queries` property) was better in this aspect. It was clear that the content under the `data_queries` property describes data queries.

**Comment by m-burgoyne** [comment]
@iandruska-ibl would it help if an attribute was added to the `variables` object to identify the type of data query rather than relying on  `title` or `description` i.e. in the example below I have added a `query_type` attribute

```json
"variables": {
    "title": "Radius query",
     "description": "Radius query",
     "query_type": "radius",
     "output_formats": [
            "CoverageJSON",
             "GeoJSON",
             "CSV"
       ],
       "default_output_format": "GeoJSON",
       "width_units": [
               "km",
                "miles"
        ],
        "crs_details": [{
             "crs": "CRS84",
             "wkt": "GEOGCS[\"WGS 84\",DATUM[\"WGS_1984\",SPHEROID[\"WGS 84\",6378137,298.257223563,AUTHORITY[\"EPSG\",\"7030\"]],AUTHORITY[\"EPSG\",\"6326\"]],PRIMEM[\"Greenwich\",0,AUTHORITY[\"EPSG\",\"8901\"]],UNIT[\"degree\",0.01745329251994328,AUTHORITY[\"EPSG\",\"9122\"]],AUTHORITY[\"EPSG\",\"4326\"]]"
          }]    
}
```

**Comment by iandruska-ibl** [comment]
@m-burgoyne Yes, that would be sufficient. Thanks.

**Comment by m-burgoyne** [comment]
@chris-little @iandruska-ibl @tomkralidis Added a query_type attribute to the variables schema and updated the examples

File: candidate-standard/openapi/examples/collection.json [hunk 3]
```diff
@@ -0,0 +1,646 @@
...
+                    "href": "http://www.example.org/edr/collections/hrly_obs/radius?coords={coords}",
+                    "hreflang": "en",
+                    "rel": "data",
+                    "title": "",
+                    "variables": {
+                        "title": "Radius query",
+                        "description": "Radius query",
+                        "query_type": "radius",
+                        "templated": true,
```
Comment (tomkralidis) [inline candidate-standard/openapi/examples/collection.json:94]: `"templated": true` should be part of the link object, not the variables object (needs to be updated in other areas of this PR as well).

**Review by tomkralidis — CHANGES_REQUESTED** [review]
We need to clarify whether or not we want our link `variables` content model to be consistent with what has been suggested in https://github.com/opengeospatial/ogcapi-environmental-data-retrieval/issues/233#issuecomment-808273775 (and OARec).

**Comment by iandruska-ibl** [comment]
The current proposal does not make much sense to me. Consider the following example from [examples/collections_metadata_JSON_1.adoc](https://github.com/opengeospatial/ogcapi-environmental-data-retrieval/blob/8070fadee4d215cb44a22ba5fc1ed0393fbe8859/candidate-standard/examples/collections_metadata_JSON_1.adoc):
```
                {
                    "href": "http://www.example.org/edr/collections/hrly_obs/position?coords={coords}",
                    "hreflang": "en",
                    "rel": "data",
                    "title": "",
                    "templated": true,
                    "variables": {
                        "title": "Position query",
                        "description": "Position query",
                        "query_type": "position",
                        "coords" :{
                            "description": "Well Known Text POINT value i.e. POINT(-120, 55)"
                        },
                        "output_formats": [
                            "CoverageJSON",
                            "GeoJSON",
                            "IWXXM"
                        ],
                        "default_output_format": "IWXXM",
                        "crs_details": [
                            {
                                "crs": "CRS84",
                                "wkt": "GEOGCS[\"WGS 84\",DATUM[\"WGS_1984\",SPHEROID[\"WGS 84\",6378137,298.257223563,AUTHORITY[\"EPSG\",\"7030\"]],AUTHORITY[\"EPSG\",\"6326\"]],PRIMEM[\"Greenwich\",0,AUTHORITY[\"EPSG\",\"8901\"]],UNIT[\"degree\",0.01745329251994328,AUTHORITY[\"EPSG\",\"9122\"]],AUTHORITY[\"EPSG\",\"4326\"]]"
                            }
                        ]
                    }
                },
```
I see two fundamentally different types of information mixed there:
1. URL query parameter description (property `coords`)
2. endpoint/data query metadata (properties: `title`, `description`, `query_type`, ...)

(1) is perfectly fine in my opinion. (2) is what bothers me. I do not think it conceptually fits to [Tom's proposal](https://github.com/opengeospatial/ogcapi-environmental-data-retrieval/issues/233#issuecomment-808273775). `title`, `description` etc. are constants describing data query, not URL query variable. It is just very confusing in my opinion. Also why only data queries are described this way? Why not collections, instances, parameters, instances as well?

I think that data query is a first-class citizen just like collection, instance or parameter and should be described that way. Currently it is just an information attached to a link. Conceptually, it is wrong because data query is not a property of a link. It is a first-class EDR concept.

Hypothetically, what if we need two different links to the same data query endpoint in the future? Do we repeat the data query metadata for each link?

My proposal is to keep (1) and transform (2) back to a property of collection/instance. For example:
```
    "data-queries": {
        "position": {
          "id": "position",
          "title": "Position Query",
          "description": "Data at point location",
          "links": [
            {
              "href": "https://ogcie.iblsoft.com/edr/collections/GFS_0-isoterm/position?coords={coords}",
              "rel": "data",
              "title": "Position Query",
              "templated": true,
              "variables": {
                 "coords" : {
                   "description": "Well Known Text POINT value i.e. POINT(-120, 55)",
                   "type": "string"
                 },
              }
            },  
          ],
          "output-formats": [
            "CoverageJSON",
            "KML",
            "NetCDF3",
            "NetCDF3_float",
            "NetCDF4",
            "NetCDF4_float"
          ],
          "default-output-format": "CoverageJSON",
          "crs-details": [
            {
              "crs": "CRS:84",
              "wkt": "GEOGCS[\"Unknown\", DATUM[\"Unknown\", SPHEROID[\"WGS_1984\", 6378137.0, 298.257223563]], PRIMEM[\"Greenwich\",0], UNIT[\"degree\", 0.017453], AXIS[\"Lon\", EAST], AXIS[\"Lat\", NORTH]]"
            }
          ]
        },
```

**Comment by m-burgoyne** [comment]
I think there is a strong argument that the OpenAPI documentation would be a better location to supply information required to generate a valid query.  A lot of the changes we are implementing here are duplicating functionality that could be provided by the OpenAPI endpoint.

**Review by tomkralidis — APPROVED** [review]

**Comment by m-burgoyne** [comment]
@tomkralidis @chris-little @iandruska-ibl, updated solution differs slightly from the example, I tried to avoid duplicating information and reuse the existing schemas.

**Review by chris-little — APPROVED** [review]
Looks good.

**Review by tomkralidis — APPROVED** [review]

**Review by iandruska-ibl — APPROVED** [review]

**Pull request merged**